File: scala/src/main/scala/problems/Problem101FindKthSmallestBst.scala

package problems
import scala.collection.mutable

object Problem101FindKthSmallestBst {
  final class Node(var v:Int, var l:Node, var r:Node)
  def kthSmallest(root: Node, k:Int): Int = {
    val st = new mutable.ArrayDeque[Node]()
    var cur: Node = root; var count = 0
    while (cur!=null || st.nonEmpty) {
      while (cur!=null) { st.append(cur); cur = cur.l }
      cur = st.removeLast()
      count += 1
      if (count==k) return cur.v
      cur = cur.r
    }
    -1
  }
}

File: java/src/main/java/problems/Problem101FindKthSmallestBst.java
package problems;


public class Problem101FindKthSmallestBst {
    public static class Node{int v; Node l,r; Node(int v){this.v=v;}}
    public static Integer kthSmallest(Node root, int k){
        java.util.Deque<Node> st=new java.util.ArrayDeque<>();
        Node cur=root; int count=0;
        while(cur!=null || !st.isEmpty()){
            while(cur!=null){ st.push(cur); cur=cur.l; }
            cur=st.pop();
            if(++count==k) return cur.v;
            cur=cur.r;
        }
        return null;
    }
}

